xbmc/pvr/providers/PVRProviders.cpp

```diff
@@ -0,0 +1,384 @@
+{
+  CSingleLock lock(m_critSection);
+  const std::shared_ptr<CPVRProvider> providerToUpdate =
+      GetByClient(provider->GetClientId(), provider->GetUniqueId());
+  if (providerToUpdate)
+  {
+    return providerToUpdate->UpdateEntry(provider, ProviderUpdateMode::BY_CLIENT);
+  }
+  else
+  {
+    provider->SetDatabaseId(++m_iLastId);
+    InsertEntry(provider, ProviderUpdateMode::BY_CLIENT);
+  }
+
+  return true;
+}
+
+std::shared_ptr<CPVRProvider> CPVRProvidersContainer::GetByClient(int iClientId,
+                                                                  int iUniqueId) const
+{
+  CSingleLock lock(m_critSection);
+  for (const auto& provider : m_providers)
+  {
+    if (provider->GetClientId() == iClientId && provider->GetUniqueId() == iUniqueId)
+    {
+      return provider;
+    }
+  }
+
+  return {};
+}
+
+void CPVRProvidersContainer::InsertEntry(const std::shared_ptr<CPVRProvider>& newProvider,
+                                         ProviderUpdateMode updateMode)
+{
+  bool found = false;
+  for (auto& provider : m_providers)
+  {
+    if (provider->GetClientId() == newProvider->GetClientId() &&
+        provider->GetUniqueId() == newProvider->GetUniqueId())
+    {
+      found = true;
+      provider->UpdateEntry(newProvider, updateMode);
+    }
+  }
+
+  if (!found)
+  {
+    m_providers.emplace_back(newProvider);
+  }
+}
+
+std::vector<std::shared_ptr<CPVRProvider>> CPVRProvidersContainer::GetProvidersList() const
+{
+  CSingleLock lock(m_critSection);
+  return m_providers;
+}
+
+
+bool CPVRProviders::Load()
+{
+  // unload previous providers
+  Unload();
+
+  // load providers from database
+  bool bReturn = LoadFromDatabase();
+
+  // update from clients
+  Update();
+
+  return bReturn;
+}
+
+void CPVRProviders::Unload()
+{
+  // remove all tags
+  CSingleLock lock(m_critSection);
+  m_providers.clear();
+}
+
+bool CPVRProviders::LoadFromDatabase()
+{
+  const std::shared_ptr<CPVRDatabase> database = CServiceBroker::GetPVRManager().GetTVDatabase();
+  if (database)
+  {
+    bool bChanged = false;
+
+    CPVRProviders providers;
+    database->Get(providers);
+
+    for (auto& provider : providers.GetProvidersList())
+    {
+      bChanged |= (CheckAndAddEntry(provider, ProviderUpdateMode::BY_DATABASE) != nullptr);
+    }
+  }
+  return true;
+}
+
+bool CPVRProviders::Update()
+{
+  {
+    CSingleLock lock(m_critSection);
+    if (m_bIsUpdating)
+      return false;
+    m_bIsUpdating = true;
+  }
+
+  // Default client providers are the add-ons themselves, we retrieve both enabled
+  // and disbled add-ons as we don't want them removed from the DB
+  CPVRProviders newAddonProviderList;
+  std::vector<int> disabledClients;
+  std::vector<CVariant> clientProviderInfos =
+      CServiceBroker::GetPVRManager().Clients()->GetClientProviderInfos();
+  CLog::LogFC(LOGDEBUG, LOGPVR, "Adding default providers, found %d PVR add-ons",
```

Comment:
Please update to the current log standards using `{}`.

For more information please see the [libfmt documentation](https://fmt.dev/latest/syntax.html)